### ISSUE TITLE
Log depth sensitivity during training — and stop the yardstick silently ignoring --depth

### ISSUE THREAD

**Issue by MatthewLacerda2** [MatthewLacerda2/TinyRefinementModel#226]
Depth inertness was discovered **at the end of a 10-day run**, by eyeballing eight sampled
prompts and then confirming on 5,153 LAMBADA examples. It could have been visible on day one.

## The measurement

Every `VAL_EVERY_OPT_STEPS`, on the batch the validation probe already loads, log:

- **argmax disagreement** — fraction of positions where `argmax(logits @ depth 2)` differs
  from `argmax(logits @ depth K)`. Zero means the refine loop cannot change a single token.
- **symmetric KL** between the depth-2 and depth-K next-token distributions — a continuous
  signal that moves before any argmax flips, so it degrades gracefully instead of sitting at
  a hard zero.

Two extra forward passes on an existing batch, no backward, at the cadence validation already
runs. Cost is in the noise.

## What it would have shown

Measured after the fact on the finished champion, 5,153 examples:

```
depth   1        2        4        8
acc     9.84%    9.82%    9.78%    9.80%      spread 0.06 pp
                                              standard error 0.41 pp
```

The spread is one seventh of one standard error. Byte-identical generated text on 4 of 8
fixed prompts at every depth. **A logged metric would have made this a day-one finding
instead of a postmortem.**

## Fix the mislabel while in here

`heldout_perplexity()` in `instruments/yardstick/eval_yardstick.py:76` takes no depth
argument — it always runs at the validation probe's fixed depth. So
`--depth 1` and `--depth 8` both emit the same held-out number, and the JSON records it
beside `"eval_depth": 1`. That is a number labelled as something it is not.

Consequence today: **we have never tested depth on our own data.** Every depth result we have
is out-of-domain LAMBADA. Given the model's register is worksheet-style educational text and
LAMBADA is narrative fiction, in-domain is precisely where depth had the best chance of
mattering, and it is the one place we did not look.

Either thread the depth through, or refuse the flag when `--no-heldout` is not set. Silently
ignoring it is the only unacceptable option.

## Acceptance

- Metric appears in `metrics.csv` and in the training log line, both arches.
- A toy model trained on a depth-*requiring* task (`experiments/depth/ablation_harness.py`,
  cumulative scan) shows clearly non-zero disagreement — proving the metric can detect depth
  working, not just report zero.
- `--depth` on the yardstick either changes the held-out number or is rejected.

## Lane

`trm/train/validation.py`, `trm/runtime/metrics.py`, `instruments/yardstick/`. CPU-testable
end to end.

**Comment by MatthewLacerda2** [comment]
## Relabelled cpu+gpu, and the sample size should be ~1000x what this issue implies

The `cpu` label was right in the lane sense — this work never blocks the card — but it made
the *measurement* sound like a CPU-scale job. It isn't, and the difference changes what the
issue can conclude.

**The binding constraint on the depth question is sample size, not compute.** Standard error
on a proportion at p≈0.098:

```
   positions         SE    2σ resolves   
       5,153    0.4142pp     0.828pp     the LAMBADA set — what we ran
      51,200    0.1314pp     0.263pp     100 forward passes
     512,000    0.0416pp     0.083pp     1k forwards, ~1 min on GPU
   5,120,000    0.0131pp     0.026pp     10k forwards, ~10 min
  51,200,000    0.0042pp     0.008pp     100k forwards, ~1.7 h
```

The observed depth spread across d1/d2/d4/d8 was **0.06 pp**. Against the LAMBADA set's
0.41 pp standard error that is 0.15σ — correctly reported as "indistinguishable from zero",
but that phrasing hides that we could not have detected a real effect of that size *even if
it existed*.

At 5M positions the same 0.06 pp would be **4.5σ**. That is the difference between "too noisy
to tell" and "we now know whether depth does exactly nothing or something very small."

## The enabler: in-domain has no sample cap

LAMBADA is fixed at 5,153 examples — that ceiling cannot be raised. But this issue already
flags that **we have never tested depth on our own data**, and `runs/data/pretrain` is 40 GB
of tokenized corpus. Every sequence yields `MAX_SEQ_LEN=512` positions, so samples are
effectively unlimited and free.

So the in-domain depth test — the one measurement we are missing — is also the one with no
sample ceiling. Run it at 5M+ positions and it becomes the tightest number in the project
rather than another suggestive one.

## Consequences for the spec

- Report the **standard error alongside every rate**, and state the smallest effect the
  chosen n could have detected. "No difference" without that second number is not a result.
- Size n from a pre-registered minimum detectable effect, not from convenience.
- The training-time metric stays cheap (it runs on the val batch every 64 steps); the
  **offline confirmation** is where the large n goes. Two different jobs, same code path.
- Sub-slice by domain (web / code / math) — with millions of positions each slice still has a
  tight error bar, and depth mattering on math but not web would be invisible in a pooled
  number.

## Not a licence to spend the card carelessly

10k forward passes is ~10 minutes on an idle 2060. 100k is 1.7h. Both are cheap, but log what
was run so a later reader knows which n produced which claim — a tight interval and a wide one
look identical once they are quoted as a single number.

**Comment by MatthewLacerda2** [comment]
## Result — depth measured in-domain for the first time

Ran the missing measurement on the idle GPU: champion checkpoint, held-out fineweb-edu (the
validation probe's slice and skip), **2,046,196 scored positions**, four depths through an
identical code path.

```
 depth        CE       ppl     argmax != d4      ±2se
     1    3.4625     31.90          0.9155%   0.0133%
     2    3.4629     31.91          0.4002%   0.0088%
     4    3.4626     31.90          0.0000%   0.0000%
     8    3.4626     31.90          0.1009%   0.0044%
```

Stable under a 10× sample increase (a 204,631-position pass gave 0.9256 / 0.4159 / 0.1065%).

## What it says, and it is not "depth does nothing"

**Depth demonstrably changes the prediction.** 0.9155% of positions differ between depth 1 and
depth 4 — **138σ** from zero. Even depth 8 vs depth 4 differs on 0.1009%, **46σ** from zero.
The refine loop is not an identity; it is active and it moves decisions.

**Depth demonstrably does not improve the prediction.** CE across a 4× compute range spans
**0.0004 nats** — 75× *below* the σ≈0.03 seed noise floor. Perplexity is 31.90 at depth 1 and
31.90 at depth 8.

Those two together bound how much a flip can be worth. If refinement gained `δ` nats on the
0.92% of tokens it changes, the pooled gain would be `0.009155 × δ ≤ 0.0004`, so:

> **|δ| ≤ 0.044 nats per changed token — and the sign is not established.**

The loop spends 4× the compute to change its mind about one token in a hundred, and the
changes are a wash: roughly as many go wrong as go right.

## This upgrades the open question rather than closing it

The LAMBADA sweep supported "depth does nothing." In-domain at 2M positions the picture is
sharper and more useful: **the mechanism works and produces nothing.** That is precisely the
hypothesis #227 exists to discriminate — is the motion *aimed* at the decision, or is it
jitter that flips near-ties by luck? A ~1% flip rate with exactly zero pooled benefit is what
unaimed jitter looks like, and #227's projection onto `e_top1 − e_top2` would confirm or kill
it directly.

It also sharpens #219 (latent noise): the premise there was partly "the block idles." It does
not idle. The problem is not that refinement fails to move the state, it is that the movement
does not carry information the readout can use.

## Caveats, stated because they bound the claim

- **The absolute CE is not comparable to the recorded val CE** (3.6473 / ppl 38.37). This
  scores 2,000 rows where the probe scores `EVAL_ROWS`, and window handling may differ. The
  *depth comparison* is unaffected — all four depths used one code path on identical data —
  but do not quote 31.90 as this run's validation number until the gap is explained.
- **fineweb-edu only.** Code and math slices are untested, and depth plausibly matters more on
  algorithmic content — the toy tasks where depth recurrence was originally proven were
  algorithmic. Sub-slicing by domain is the obvious next cut.
- One checkpoint. Running the ladder (0.78B / 1.40B / 3.20B) would show whether the flip rate
  is growing or shrinking over training, which is a different and more interesting question
  than its endpoint value.
- The script's printed "smallest 2σ-detectable rate" uses p=0.5, the worst case, so it is
  over-conservative for rare events; the per-depth `±2se` columns are the ones to read.

Script: `aux_depth/indomain_depth.py` (gitignored scratch). ~3,150 positions/sec on the 2060;
the 2M-position run took ~11 minutes. Cheap enough to re-run per checkpoint and per domain.

**Comment by MatthewLacerda2** [comment]
## Correction and completion — the paired test, across all three domains

**First, a correction to my previous comment.** I compared depth-to-depth cross-entropy
against the σ≈0.03 seed noise floor (#17). That is the wrong bar. #17 measures *run-to-run*
variance across training seeds; this is **one set of weights scored twice on identical
tokens**, so there is no seed variance to clear. The right test is paired — per position,
`CE(depth d) − CE(depth 4)` — where each token's difficulty cancels and only the depth
effect survives. Redone properly below. The earlier conclusion changes in an interesting way.

## Paired result — 2.0M positions per domain, reference depth 4

Negative = that depth has *lower* CE than depth 4.

```
                  depth    mean Δce    paired se       t
codeparrot            1   +0.001259     0.000057   +22.3   worse
                      2   -0.000128     0.000034    -3.8   better
                      8   -0.000213     0.000021    -9.9   better

fineweb-edu           1   -0.000119     0.000017    -7.2   better   <- d1 BEATS d4
                      2   +0.000233     0.000008   +30.6   worse
                      8   -0.000006     0.000002    -3.2   better

finemath              1   -0.000056     0.000014    -3.9   better   <- d1 BEATS d4
                      2   +0.000156     0.000007   +22.3   worse
                      8   -0.000009     0.000002    -4.4   better
```

**Depth helps on code, and only on code.** `codeparrot` is the one domain with a coherent
ordering — depth 1 is clearly the worst, and more depth is better. That is the shape the
architecture predicts, and it is consistent with depth recurrence having originally been
proven on *algorithmic* toy tasks.

**On web and math the ordering is scrambled, and depth 1 beats depth 4 on both.** Whatever
the loop is doing there, it is not refining toward a better answer.

Note the identical sign pattern on fineweb-edu and finemath — d1 better, d2 distinctly worse
(t = +30.6 and +22.3), d8 marginally better. Two independent corpora producing the same
peculiar shape looks more like a systematic artifact than a depth effect, and is worth a look
on its own. (Speculation, untested: the time-signal rows are trained at very unequal
frequency, since depth is drawn uniformly from 1..8 and row 0 appears in every draw while
row 7 appears in one in eight.)

## Significance is not the story — effect size is

At n = 2,000,000 a t of 22 certifies an effect that is economically meaningless. The largest
depth benefit measurable anywhere is codeparrot's d1→d4 at **0.00126 nats**. Calibrated:

```
  seed noise floor (#17)                                 0.0300 nats     24x larger
  the last third of the 4B run, already called negligible 0.0170 nats     13x larger
  total CE gained across the whole 4B run                 4.8360 nats   3838x larger
```

And the cost, corrected — I previously said "4× the compute", which is wrong. The encoder is
7 distinct blocks and only the refine block loops, so per-forward cost in block-equivalents:

```
  depth 1:  8 blocks   1.000x
  depth 2:  9 blocks   1.125x
  depth 4: 11 blocks   1.375x
  depth 8: 15 blocks   1.875x
```

So the best case is **+37.5% forward compute for 0.00126 nats on one domain of three, and a
small loss on the other two.**

## Where this leaves the depth question

Refinement is **real, measurable, domain-dependent, and economically worthless at this scale.**
That is a sharper claim than either "depth does nothing" (my LAMBADA reading) or "the block
idles" (the premise I gave #219, now retracted twice over — it neither idles nor is inert).

The live hypotheses this leaves, in the order I would test them:

1. **Code is a hint, not a result.** One domain out of three, at 0.001 nats. Re-run across the
   0.78B / 1.40B / 3.20B ladder: if the code advantage is *growing* with training, depth may be
   a mechanism that needs more scale to matter. If flat or shrinking, it is a curiosity.
2. **#227 becomes the priority.** A coherent gain on code and scrambled noise on prose is
   exactly what "aimed on some content, jitter on the rest" would look like. The
   `e_top1 − e_top2` projection would separate those directly, per domain.
3. **The d2 anomaly** — worse than both d1 and d4 on two independent corpora — is unexplained
   and cheap to investigate.

Scripts (gitignored scratch): `aux_depth/indomain_depth.py` (pooled + argmax disagreement),
`aux_depth/paired_depth.py` (this test). ~3,150 positions/sec on the 2060; ~10 min per domain.
Both should be folded into the instrument this issue describes rather than left in `aux_`.

**Comment by MatthewLacerda2** [comment]
## Why depth helps on code: it is tracking hierarchical structure

The pooled result ("depth helps on codeparrot, not on fineweb-edu or finemath") hides where
the gain sits. Breaking `CE(depth 1) − CE(depth 8)` down per position, champion checkpoint,
~410k positions per corpus, answers it. Positive = depth 8 better.

### The gain is delimiters, and almost nothing else

```
codeparrot — pooled +0.001827 nats
  category            mean gain      % of tokens   % of total gain
  bracket             +0.014196            5.9%            45.7%
  punctuation         +0.004151           13.3%            30.3%
  whitespace          +0.001262           40.4%            27.9%
  number              -0.000496            1.7%            -0.5%
  word/identifier     -0.000378           38.4%            -8.0%
```

**Delimiters are 19.2% of tokens and carry 76% of the gain.** Brackets alone: 5.9% of tokens,
45.7% of the gain.

### The prose control says it is not the character, it is the job

```
fineweb-edu — pooled -0.000140 nats (i.e. nothing)
  bracket             +0.000358      <- 40x smaller than the same token in code
  punctuation         -0.001926
  word/identifier     +0.000206
```

A bracket in prose is a parenthetical aside; a bracket in code opens a scope that must be
tracked and closed. Same token, same tokenizer, same model — and the gain only appears where
the delimiter carries structure.

(Share-of-gain percentages are deliberately omitted for fineweb-edu: its pooled total is
≈0, so dividing by it produces meaningless figures like "-119%". Only the raw means are
interpretable there.)

### Nesting depth, in code

```
  nest 0   +0.000861   (255,582 positions)
  nest 1   +0.003601   (108,631)
  nest 2   +0.003310    (33,012)
  nest 3   +0.001166     (9,197)
  nest 4+  +0.005405     (3,054)
```

Roughly 4x more gain inside a bracket structure than at top level. In prose the same
breakdown is flat-to-negative at every nesting level.

## What this kills

- **Tokenizer artifact — dead.** Whitespace is 40.4% of code tokens and returns only 27.9% of
  the gain: *under*-represented relative to its share. Had r50k's whitespace handling been the
  story, this is where it would have shown.
- **Long-range exact copying — dead.** `word/identifier` gain is **negative** (−0.000378), and
  tokens seen earlier in the window gain *less* (+0.001427) than first-time tokens
  (+0.003544). If depth were doing retrieval-and-copy, identifiers would be the winners. They
  are the losers.
- **"Depth resolves uncertainty" — dead.** Gain rises with the model's confidence, peaking at
  p 0.8–0.9 (+0.003986) and falling at both tails. Depth is not rescuing hard calls; it is
  sharpening ones the model is already fairly sure of.

## What survives

**Depth is doing state tracking over nested structure** — maintaining what is open and what
must close. That is a cumulative, sequential computation, and it is *exactly* the class of
task depth recurrence was proven on here originally
(`docs/findings/2026-06-13-plan-a-depth-recurrence-works.md`, cumulative-scan toy tasks).

The mechanism appears to work as designed. It contributes almost nothing to a language-model
loss because **prose has very little of the structure it is good at**, and prose is most of
what we train on.

That reframes the architecture question. It is not "does depth recurrence work" — on this
evidence it does, on the thing it was built for. It is **"is nested-structure tracking worth
1.4x forward compute for a general LM"**, and at this scale the pooled answer is no:
+0.0018 nats on the one corpus where it applies.

## Caveats

- One checkpoint. The ladder (#226) shows depth *hurting* at 0.78B and 1.40B, so this
  attribution should be re-run across checkpoints — if the delimiter gain grows with training,
  that is a different and much more interesting story than its endpoint value.
- `finemath` not yet broken down. It showed no pooled gain, but it is worth checking whether
  it has a delimiter signal drowned by prose-like tokens — LaTeX and equations have their own
  bracket structure.
- Categories come from decoding single tokens with `r50k_base`; a token spanning a bracket and
  a letter lands in `bracket`. The `mixed` bucket (n=1,024, +0.033) is small and noisy.
- Nesting is counted from decoded brackets only, so it ignores Python's indentation-based
  scoping — meaning the nesting effect here is probably *understated* for Python.

## Status

I would treat this as a **findings-file candidate** rather than a finished finding: I could not
locate prior work attributing depth-recurrence gains to specific token classes, which by the
uncertain→novel rule makes it novel. But it needs the cross-checkpoint replication above
before it earns a dated entry. Script: `aux_depth/why_code.py`.

**Comment by MatthewLacerda2** [comment]
superseded-by #247 — depth recurrence was retired as the live bet on 2026-09-12 and `MODEL_ARCH` now defaults to `plain`, which has no refine loop, no depth dial and no trajectory. This issue's subject no longer exists on the architecture we train.

Evidence: `docs/findings/2026-09-12-depth-recurrence-is-suppressed-not-exploited.md`.

`MODEL_ARCH=refiner` is still selectable (it is the 4B champion's architecture), so nothing here is *unrunnable* — it is just no longer measuring the live path, which is what made it worth doing.

What survives from it is folded into the activation-telemetry work: logging per-block `max|activation|` during a run is the arch-independent version of "watch what the stack does while it trains", and it is the check that would have caught #235 while the run was happening rather than two weeks after it ended.
